src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -285,54 +285,164 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
+                            result = Sse.Add(result, x2);
+                            ppos++;
+                        }
+
+                        Sse.Store(pDstCurrent, result);
+                        pDstCurrent += misalignment;
+                        pm0 += misalignment * ccol;
+                        length -= misalignment;
+                    }
+
+                    if (length > 3)
                     {
-                        int col = *ppos;
-                        Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
-                        Vector128<float> x2 = Sse.SetAllVector128(pSrcCurrent[col]);
-                        x2 = Sse.Multiply(x2, x1);
-                        result = Sse.Add(result, x2);
+                        remainder = length % 4;
+                        while (pDstCurrent < pDstEnd)
+                        {
+                            float* pm1 = pm0 + ccol;
+                            float* pm2 = pm1 + ccol;
+                            float* pm3 = pm2 + ccol;
+                            Vector128<float> result = Sse.SetZeroVector128();
+
+                            int* ppos = pposMin;
 
-                        ppos++;
+                            while (ppos < pposEnd)
+                            {
+                                int col = *ppos;
+                                Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
+                                Vector128<float> x2 = Sse.SetAllVector128(pSrcCurrent[col]);
+                                x2 = Sse.Multiply(x2, x1);
+                                result = Sse.Add(result, x2);
+                                ppos++;
+                            }
+
+                            Sse.Store(pDstCurrent, result);
+                            pDstCurrent += 4;
+                            pm0 += 4 * ccol;
+                        }
+                    }
+                    else
+                    {
+                        remainder = length;
                     }
 
-                    Sse.StoreAligned(pDstCurrent, result);
-                    pDstCurrent += 4;
-                    pm0 += 4 * ccol;
+                    if (remainder != 0)
+                    {
+                        pDstCurrent -= (4 - remainder);
+                        pm0 -= (4 - remainder) * ccol;
+                        Vector128<float> trailingMask = Sse.LoadVector128(((float*)(pTrailingAlignmentMask)) + (remainder * 4));
+                        Vector128<float> leadingMask = Sse.LoadVector128(((float*)(pLeadingAlignmentMask)) + ((4 - remainder) * 4));
+
+                        float* pm1 = pm0 + ccol;
+                        float* pm2 = pm1 + ccol;
+                        float* pm3 = pm2 + ccol;
+                        Vector128<float> result = Sse.SetZeroVector128();
+
+                        int* ppos = pposMin;
+
+                        while (ppos < pposEnd)
```

Comment:
Can we refactor this inner loop into its own method? It is duplicated 3 times in this method.